### PR TITLE
refactor: split reth-stages into reth-stages-api and reth-stages

### LINKED ISSUE #7599: Split `reth-stages` into `reth-stages-api` and `reth-stages`
### Describe the feature

Split the `reth-stages` crate into two: one that just contains the traits and common types (e.g. `ExecInput`, checkpoints) and one that contains the implementations of the stages in Reth.

This split should make the dependency tree for crates that only depend on `reth-stages` for the traits/common types smaller, improving compilation times and reducing the chance that we encounter a circular dependency

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by AbnerZheng**
close https://github.com/paradigmxyz/reth/issues/7599

### CODE REVIEW

**Comment by AbnerZheng** [comment]
@onbjerg PTAL whether it is in the right direction, and if so, is there any more common trait should be extract int reth-stage-api?

File: bin/reth/src/commands/stage/dump/execution.rs [hunk 1]
```diff
@@ -10,7 +10,8 @@ use reth_node_ethereum::EthEvmConfig;
 use reth_primitives::stage::StageCheckpoint;
 use reth_provider::{ChainSpecProvider, ProviderFactory};
 use reth_revm::EvmProcessorFactory;
-use reth_stages::{stages::ExecutionStage, Stage, UnwindInput};
+use reth_stages::stages::ExecutionStage;
+use reth_stages_api::{Stage, UnwindInput};
```
Comment (mattsse) [inline bin/reth/src/commands/stage/dump/execution.rs:14]: I'd like to reexport api types from stages, this way we don't need to change anything here

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/stages/src/pipeline/set.rs [hunk 2]
```diff
@@ -1,6 +1,6 @@
-use crate::Stage;
 use reth_db::database::Database;
 use reth_primitives::stage::StageId;
+use reth_stages_api::Stage;
```
Comment (onbjerg) [inline crates/stages/src/pipeline/set.rs:3]: this should be moved to the api crate

**Review by onbjerg — CHANGES_REQUESTED** [review]
lgtm overall but yes, please re-export the api types from the reth-stages crate to simplify the dependencies for crates that require types from both

i'd also like to move the pipeline to the api crate if possible, since this should not depend on any stage impls

i.e. reth-stages is only the impl of the various stages + the default stage sets defines in crates/stages/src/sets.rs

File: bin/reth/src/commands/stage/dump/execution.rs (hunk 1 above)
Comment (AbnerZheng) [reply bin/reth/src/commands/stage/dump/execution.rs]: Done.

File: crates/stages/src/pipeline/set.rs (hunk 2 above)
Comment (AbnerZheng) [reply crates/stages/src/pipeline/set.rs]: How about "test_utils"? Is it good idea to move them into "stages_api"?

Comment (mattsse) [reply crates/stages/src/pipeline/set.rs]: imo test-utils should stay in reth-stages

**Comment by mattsse** [comment]
sorry about the conflicts -.-

**Comment by AbnerZheng** [comment]
> sorry about the conflicts -.-

No worries

File: crates/stages/src/pipeline/set.rs (hunk 2 above)
Comment (AbnerZheng) [reply crates/stages/src/pipeline/set.rs]: It makes sense. But the tests of `pipeline` still relay on `reth_stages::util_utils::TestStage`, and while `pipeline` has been moved into `reth_stages_api`, I moved `reth_stages::util_utils::TestStage` into `reth_stages_api::util_utils::TestStage`, is there any problem?

**Review by onbjerg — COMMENTED** [review]
this looks good to me overall now, will give a final review once we're out of draft. thanks for taking this on 😄

**Comment by onbjerg** [comment]
Hey @AbnerZheng, thanks for starting this! I went ahead and applied a few fixes to get it over the line since I need this 😄

**Review by onbjerg — APPROVED** [review]

**Review by mattsse — APPROVED** [review]
lgtm

perhaps we should consider grouping the stages crates in a new folder `sync/{stages-api,stages}`?

**Comment by AbnerZheng** [comment]
> Hey @AbnerZheng, thanks for starting this! I went ahead and applied a few fixes to get it over the line since I need this 😄

Wait a moment, let me remove TODOs I added.

**Comment by AbnerZheng** [comment]
> > Hey @AbnerZheng, thanks for starting this! I went ahead and applied a few fixes to get it over the line since I need this 😄
> 
> Wait a moment, let me remove TODOs I added.

Will fire a PR to remove the unnecessary TODO I added soon.

**Comment by onbjerg** [comment]
Ah, sorry 🙈

**Pull request merged**